File: benchmark-tablets/pipette-clients/ios/Pipette/Pipette/Networking/ManagementClient.swift
import CryptoKit
import Foundation

enum ManagementClient {
    struct RegistrationResult {
        let clientId: ClientID
        let status: String
    }

    private static let userAgent = "pipette-ios"
    /// Empty JSON object body for POST/PUT endpoints that require a body.
    private static let emptyJSONObject = Data("{}".utf8)

    /// `POST /clients/register`. **Unauthenticated.**
    ///
    /// `device` carries the same profile + capability fields the server accepts
    /// on `PATCH /clients/me`; supplying them here establishes accurate matching
    /// input in one request, so a first-time client needs no follow-up patch
    /// before it is matchable (mgmt `client-integration.md` §2).
    ///
    /// Required rather than defaulted: registering with no matching input is the
    /// state this parameter exists to prevent, so there is no sensible default to
    /// fall back to. Pass an explicitly empty `DeviceProfileUpdate()` to opt out.
    static func register(
        serverUrl: ServerURL,
        organization: String,
        contactEmail: String,
        clientDetails: String,
        publicKeyHex: String,
        preauthKey: String? = nil,
        device: DeviceProfileUpdate
    ) async throws -> RegistrationResult {
        let body = RegisterRequest(
            publicKey: publicKeyHex,
            organization: organization,
            contactEmail: contactEmail,
            clientDetails: clientDetails,
            preauthKey: preauthKey,
            device: device
        )
        let response: RegisterResponse = try await request(
            serverUrl: serverUrl,
            path: "/clients/register",
            method: "POST",
            auth: nil,
            body: try Coding.encoder.encode(body)
        )
        return RegistrationResult(
            clientId: ClientID(response.clientId),
            status: response.status
        )
    }

    static func submitResult(
        serverUrl: ServerURL,
        auth: AuthIdentity,
        payloadJson: String
    ) async throws -> String {
        guard let body = payloadJson.data(using: .utf8) else {
            throw ManagementClientError.invalidPayloadEncoding
        }

        let responseData = try await requestData(
            serverUrl: serverUrl,
            path: "/benchmarks",
            method: "POST",
            auth: auth,
            body: body
        )
        guard let response = String(data: responseData, encoding: .utf8) else {
            throw ManagementClientError.invalidResponseEncoding
        }
        return response
    }

    // MARK: - Planner (claim / heartbeat / reclaim / profile)

    /// `POST /plans/claim` — lease the next eligible job, or `nil` on `204`.
    /// Throws on `403` (not approved) and other errors.
    static func claim(
        serverUrl: ServerURL,
        auth: AuthIdentity
    ) async throws -> ClaimedJob? {
        let response = try await requestResponse(
            serverUrl: serverUrl,
            path: "/plans/claim",
            method: "POST",
            auth: auth,
            body: emptyJSONObject
        )
        if response.statusCode == 204 { return nil }
        guard (200..<300).contains(response.statusCode) else {
            throw ManagementClientError.httpStatus(
                statusCode: response.statusCode,
                body: String(data: response.data, encoding: .utf8) ?? ""
            )
        }
        do {
            return try Coding.decoder.decode(ClaimedJob.self, from: response.data)
        } catch {
            throw ManagementClientError.decodeJson(error)
        }
    }

    /// `PUT /plans/{jobId}/heartbeat` — renew the lease. Empty body on success.
    static func heartbeat(
        serverUrl: ServerURL,
        auth: AuthIdentity,
        jobId: String
    ) async throws {
        try await requestEmpty(
            serverUrl: serverUrl,
            path: "/plans/\(jobId)/heartbeat",
            method: "PUT",
            auth: auth,
            body: emptyJSONObject
        )
    }

    /// `POST /plans/{jobId}/reclaim` — re-acquire a previously held lease.
    static func reclaim(
        serverUrl: ServerURL,
        auth: AuthIdentity,
        jobId: String
    ) async throws {
        try await requestEmpty(
            serverUrl: serverUrl,
            path: "/plans/\(jobId)/reclaim",
            method: "POST",
            auth: auth,
            body: emptyJSONObject
        )
    }

    /// `GET /clients/me` — authenticated client profile.
    static func me(
        serverUrl: ServerURL,
        auth: AuthIdentity
    ) async throws -> ClientProfile {
        try await request(
            serverUrl: serverUrl,
            path: "/clients/me",
            method: "GET",
            auth: auth,
            body: nil
        )
    }

    /// `PATCH /clients/me` — refresh device profile / capabilities.
    static func updateMe(
        serverUrl: ServerURL,
        auth: AuthIdentity,
        update: DeviceProfileUpdate
    ) async throws -> ClientProfile {
        try await request(
            serverUrl: serverUrl,
            path: "/clients/me",
            method: "PATCH",
            auth: auth,
            body: try Coding.encoder.encode(update)
        )
    }

    /// Submit a plan-attached failure (`message_type: "failure"`).
    static func submitFailure(
        serverUrl: ServerURL,
        auth: AuthIdentity,
        failure: FailureSubmission
    ) async throws {
        _ = try await requestData(
            serverUrl: serverUrl,
            path: "/benchmarks",
            method: "POST",
            auth: auth,
            body: try Coding.encoder.encode(failure)
        )
    }

    /// Submit an already-serialized plan-attached success payload (must carry `job_id`).
    static func submitPlanResult(
        serverUrl: ServerURL,
        auth: AuthIdentity,
        payloadJson: Data
    ) async throws {
        _ = try await requestData(
            serverUrl: serverUrl,
            path: "/benchmarks",
            method: "POST",
            auth: auth,
            body: payloadJson
        )
    }

    static func submitResultBatch(
        serverUrl: ServerURL,
        auth: AuthIdentity,
        payloadsJson: String
    ) async throws -> String {
        // Parse the pre-serialized payload array and rebuild the envelope with
        // JSONSerialization rather than string-concatenating into the JSON: a
        // concat silently produces malformed bodies if the blob is ever empty
        // or not the array the wire shape assumes. Wire shape is unchanged:
        // {"submissions": [...]}.
        guard let payloadsData = payloadsJson.data(using: .utf8),
              let submissions = try? JSONSerialization.jsonObject(with: payloadsData) as? [Any]
        else {
            throw ManagementClientError.invalidPayloadEncoding
        }
        let body = try JSONSerialization.data(withJSONObject: ["submissions": submissions])

        let responseData = try await requestData(
            serverUrl: serverUrl,
            path: "/benchmarks/batch",
            method: "POST",
            auth: auth,
            body: body
        )
        guard let response = String(data: responseData, encoding: .utf8) else {
            throw ManagementClientError.invalidResponseEncoding
        }
        return response
    }

    /// Outcome of a conditional `GET` (`If-None-Match`). `json` is the raw response
    /// body on `200`, `nil` on `304` (caller keeps its cached copy); `etag` is the
    /// server's content hash to echo back as `If-None-Match` next time.
    struct ConditionalGet {
        let json: Data?
        let etag: String?

        var notModified: Bool { json == nil }
    }

    /// Fetch the benchmark catalog (definitions only — no eval samples),
    /// ETag-conditional. The `GET /benchmarks` endpoints are public — no client id
    /// or signature — so this needs no registration. See `conditionalGet`.
    static func fetchBenchmarks(
        serverUrl: ServerURL,
        ifNoneMatch: String?
    ) async throws -> ConditionalGet {
        try await conditionalGet(serverUrl: serverUrl, path: "/benchmarks", ifNoneMatch: ifNoneMatch)
    }

    /// Fetch a single benchmark by id, ETag-conditional. Unlike the list, the
    /// per-id response includes the eval `samples`, so every benchmark must be
    /// fetched individually to be fully hydrated. Public — no auth. See
    /// `conditionalGet`.
    static func fetchBenchmark(
        serverUrl: ServerURL,
        benchmarkId: String,
        ifNoneMatch: String?
    ) async throws -> ConditionalGet {
        try await conditionalGet(
            serverUrl: serverUrl, path: "/benchmarks/\(benchmarkId)", ifNoneMatch: ifNoneMatch)
    }

    /// Unauthenticated `GET` that honors `If-None-Match`: sends the last-seen `etag`,
    /// and a `304` returns `notModified` (no body). Used for the public benchmark
    /// endpoints. (`requestData` can't be reused — it treats the `304` as a
    /// non-success status and throws.)
    private static func conditionalGet(
        serverUrl: ServerURL,
        path: String,
        ifNoneMatch: String?
    ) async throws -> ConditionalGet {
        let url = try endpoint(serverUrl: serverUrl, path: path).url
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue(userAgent, forHTTPHeaderField: "User-Agent")
        if let ifNoneMatch {
            request.setValue(ifNoneMatch, forHTTPHeaderField: "If-None-Match")
        }

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ManagementClientError.invalidResponse
        }
        let etag = http.value(forHTTPHeaderField: "ETag")
        if http.statusCode == 304 {
            return ConditionalGet(json: nil, etag: etag)
        }
        guard (200..<300).contains(http.statusCode) else {
            throw ManagementClientError.httpStatus(
                statusCode: http.statusCode,
                body: String(data: data, encoding: .utf8) ?? ""
            )
        }
        return ConditionalGet(json: data, etag: etag)
    }

    /// The signed management-auth headers for one request.
    ///
    /// `pathAndQuery` must be the request target the server receives — the base
    /// URL's own path prefix included, and the query string when there is one.
    /// The server verifies against its own `uri.path_and_query()`, so anything
    /// else fails to verify. `Endpoint` carries the right value.
    static func signedHeaders(
        auth: AuthIdentity,
        method: String,
        pathAndQuery: String,
        timestamp: String = Self.rfc3339Timestamp(),
        nonce: String = Self.generateNonce()
    ) throws -> [String: String] {
        let privateKeyData = try Data(hexEncoded: auth.privateKeyHex.value)
        let privateKey = try Curve25519.Signing.PrivateKey(rawRepresentation: privateKeyData)
        let payload = signedPayload(
            method: method,
            pathAndQuery: pathAndQuery,
            timestamp: timestamp,
            clientId: auth.clientId.value,
            nonce: nonce
        )
        let signature = try privateKey
            .signature(for: Data(payload.utf8))
            .hexEncodedString()
        return [
            "X-Client-Id": auth.clientId.value,
            "X-Timestamp": timestamp,
            "X-Nonce": nonce,
            "X-Signature": signature
        ]
    }

    /// The `v1` signed payload: six newline-separated fields — scheme tag,
    /// method, request target, timestamp, client id, nonce (mgmt
    /// `authentication.md` §2.1). Binding the method and target scopes a
    /// signature to that method and target; the nonce makes it single-use, so a
    /// captured signature cannot be replayed inside the freshness window. The
    /// request body is still not covered.
    static func signedPayload(
        method: String,
        pathAndQuery: String,
        timestamp: String,
        clientId: String,
        nonce: String
    ) -> String {
        "v1\n\(method)\n\(pathAndQuery)\n\(timestamp)\n\(clientId)\n\(nonce)"
    }

    /// A fresh per-request nonce: 16 CSPRNG bytes, lowercase hex.
    ///
    /// Hex rather than an arbitrary byte string on purpose. The nonce is a field
    /// in a newline-delimited payload, so a value carrying a newline could forge
    /// a field boundary and make two different requests hash to one payload; hex
    /// cannot. It also satisfies the server's non-empty and valid-UTF-8 rules by
    /// construction. 128 bits makes a collision across the fleet negligible, so
    /// the server's replay cache can reject a repeat without the client
    /// coordinating.
    ///
    /// `SystemRandomNumberGenerator`, which `random(in:)` uses, is documented as
    /// cryptographically secure on Apple platforms.
    static func generateNonce() -> String {
        Data((0..<16).map { _ in UInt8.random(in: .min ... .max) }).hexEncodedString()
    }

    private static func request<T: Decodable>(
        serverUrl: ServerURL,
        path: String,
        method: String,
        auth: AuthIdentity?,
        body: Data?
    ) async throws -> T {
        let data = try await requestData(
            serverUrl: serverUrl,
            path: path,
            method: method,
            auth: auth,
            body: body
        )
        do {
            return try Coding.decoder.decode(T.self, from: data)
        } catch {
            throw ManagementClientError.decodeJson(error)
        }
    }

    private struct HTTPResponse {
        let statusCode: Int
        let data: Data
    }

    private static func requestResponse(
        serverUrl: ServerURL,
        path: String,
        method: String,
        auth: AuthIdentity?,
        body: Data?
    ) async throws -> HTTPResponse {
        let endpoint = try endpoint(serverUrl: serverUrl, path: path)
        var request = URLRequest(url: endpoint.url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue(userAgent, forHTTPHeaderField: "User-Agent")

        if let body {
            request.httpBody = body
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }

        if let auth {
            let headers = try signedHeaders(
                auth: auth,
                method: method,
                pathAndQuery: endpoint.pathAndQuery
            )
            for (name, value) in headers {
                request.setValue(value, forHTTPHeaderField: name)
            }
        }

        let (data, response) = try await URLSession.shared.data(
            for: request,
            delegate: Self.refuseRedirects
        )
        guard let http = response as? HTTPURLResponse else {
            throw ManagementClientError.invalidResponse
        }
        return HTTPResponse(statusCode: http.statusCode, data: data)
    }

    /// Stateless, so one instance serves every request.
    private static let refuseRedirects = RefuseRedirects()

    private static func requestEmpty(
        serverUrl: ServerURL,
        path: String,
        method: String,
        auth: AuthIdentity?,
        body: Data?
    ) async throws {
        let response = try await requestResponse(
            serverUrl: serverUrl,
            path: path,
            method: method,
            auth: auth,
            body: body
        )
        guard (200..<300).contains(response.statusCode) else {
            throw ManagementClientError.httpStatus(
                statusCode: response.statusCode,
                body: String(data: response.data, encoding: .utf8) ?? ""
            )
        }
    }

    private static func requestData(
        serverUrl: ServerURL,
        path: String,
        method: String,
        auth: AuthIdentity?,
        body: Data?
    ) async throws -> Data {
        let response = try await requestResponse(
            serverUrl: serverUrl,
            path: path,
            method: method,
            auth: auth,
            body: body
        )
        guard (200..<300).contains(response.statusCode) else {
            throw ManagementClientError.httpStatus(
                statusCode: response.statusCode,
                body: String(data: response.data, encoding: .utf8) ?? ""
            )
        }
        return response.data
    }

    /// A resolved endpoint: the URL to send, and the request target to sign.
    struct Endpoint {
        let url: URL
        /// The path the server receives — the base URL's own prefix included —
        /// plus the query string when there is one, percent-encoded exactly as
        /// it goes on the wire. This is what `signedHeaders` must cover.
        let pathAndQuery: String
    }

    static func endpoint(serverUrl: ServerURL, path: String) throws -> Endpoint {
        let trimmed = serverUrl.value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard var components = URLComponents(string: trimmed),
              components.scheme != nil,
              components.host != nil
        else {
            throw ManagementClientError.invalidBaseURL(serverUrl.value)
        }

        let basePath = components.path.trimmingTrailingSlashes()
        components.path = basePath + path

        guard let url = components.url else {
            throw ManagementClientError.invalidBaseURL(serverUrl.value)
        }
        // Read back from the same components the URL was built from, so the
        // signed target and the sent target cannot drift.
        let encodedPath = components.percentEncodedPath
        let pathAndQuery = components.percentEncodedQuery.map { "\(encodedPath)?\($0)" } ?? encodedPath
        return Endpoint(url: url, pathAndQuery: pathAndQuery)
    }

    private static func rfc3339Timestamp() -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        return formatter.string(from: Date())
    }
}

/// Task delegate that refuses redirects, so the `3xx` surfaces as the response
/// instead of being followed.
///
/// The `v1` signature covers the request target: following a redirect would
/// present a signature over the pre-redirect target and earn a `401`, and
/// `URLSession` forwards custom headers, `X-Signature` included, to the redirect
/// host.
///
/// Implements the completion-handler form rather than the `async` one: the
/// `async` variant makes SILGen crash emitting the method's ObjC thunk
/// (`emitObjCMethodThunk`) in a debug build.
private final class RefuseRedirects: NSObject, URLSessionTaskDelegate {
    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest request: URLRequest,
        completionHandler: @escaping (URLRequest?) -> Void
    ) {
        completionHandler(nil)
    }
}

enum ManagementClientError: LocalizedError {
    case invalidBaseURL(String)
    case invalidPayloadEncoding
    case invalidPrivateKeyHex
    case invalidResponse
    case invalidResponseEncoding
    case httpStatus(statusCode: Int, body: String)
    case decodeJson(Error)

    var errorDescription: String? {
        switch self {
        case .invalidBaseURL(let value):
            return "Invalid management server URL: \(value)"
        case .invalidPayloadEncoding:
            return "Failed to encode result payload"
        case .invalidPrivateKeyHex:
            return "Invalid private key"
        case .invalidResponse:
            return "Invalid response from management server"
        case .invalidResponseEncoding:
            return "Management server returned non-UTF-8 response data"
        case .httpStatus(let statusCode, let body):
            if body.isEmpty {
                return "Management server returned HTTP \(statusCode)"
            }
            return "Management server returned HTTP \(statusCode): \(body)"
        case .decodeJson(let error):
            return "Failed to parse management server response: \(error.localizedDescription)"
        }
    }
}

// Internal (not private) so a test can pin the wire contract by encoding it
// directly; the app can only build one via `ManagementClient.register`.
struct RegisterRequest: Encodable {
    let publicKey: String
    let organization: String
    let contactEmail: String
    let clientDetails: String
    /// Optional pre-auth key that admits the client already `approved`. Omitted
    /// from the JSON when nil, so keyless registration's wire body carries no
    /// empty key. Transient — the secret is never persisted or logged.
    let preauthKey: String?
    /// Device profile + capabilities, encoded **flat** alongside the fields
    /// above rather than nested under a key of their own. Empty by default, in
    /// which case it contributes nothing and the wire body is byte-for-byte the
    /// original five-field shape.
    let device: DeviceProfileUpdate

    enum CodingKeys: String, CodingKey {
        case publicKey = "public_key"
        case organization
        case contactEmail = "contact_email"
        case clientDetails = "client_details"
        case preauthKey = "preauth_key"
    }

    /// Hand-written because Swift has no `#[serde(flatten)]`: passing `encoder`
    /// straight to `device` lets it write its own snake_case keys into *this*
    /// keyed container, producing the flat object the server expects. Anything
    /// added to `DeviceProfileUpdate` therefore reaches registration for free.
    ///
    /// `client_details` is the one key both halves can write, and the flattened
    /// half goes last — so its value would silently win. Registration's own
    /// argument owns that key, so the profile's copy is dropped here rather than
    /// at the call site: a keyed container has no duplicate-key error to catch,
    /// so getting this wrong would be invisible on the wire.
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(publicKey, forKey: .publicKey)
        try container.encode(organization, forKey: .organization)
        try container.encode(contactEmail, forKey: .contactEmail)
        try container.encode(clientDetails, forKey: .clientDetails)
        try container.encodeIfPresent(preauthKey, forKey: .preauthKey)
        var device = device
        device.clientDetails = nil
        try device.encode(to: encoder)
    }
}

private struct RegisterResponse: Decodable {
    let clientId: String
    let status: String

    enum CodingKeys: String, CodingKey {
        case clientId = "client_id"
        case status
    }
}

private extension Data {
    init(hexEncoded hex: String) throws {
        let trimmed = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count.isMultiple(of: 2) else {
            throw ManagementClientError.invalidPrivateKeyHex
        }

        var bytes: [UInt8] = []
        bytes.reserveCapacity(trimmed.count / 2)

        var index = trimmed.startIndex
        while index < trimmed.endIndex {
            let next = trimmed.index(index, offsetBy: 2)
            guard let byte = UInt8(trimmed[index..<next], radix: 16) else {
                throw ManagementClientError.invalidPrivateKeyHex
            }
            bytes.append(byte)
            index = next
        }
        self.init(bytes)
    }

    func hexEncodedString() -> String {
        map { String(format: "%02x", $0) }.joined()
    }
}

private extension String {
    func trimmingTrailingSlashes() -> String {
        var result = self
        while result.last == "/" {
            result.removeLast()
        }
        return result
    }
}

File: benchmark-tablets/pipette-clients/android/Pipette/app/src/main/java/ai/liquid/pipette/ManagementClient.kt
package ai.liquid.pipette

import android.net.Uri
import java.net.HttpURLConnection
import java.net.URI
import java.net.URISyntaxException
import java.net.URL
import java.security.SecureRandom
import java.time.Instant
import java.time.format.DateTimeFormatter
import org.json.JSONArray
import org.json.JSONObject

/**
 * Thrown for a non-2xx management-server response; carries the HTTP status so callers can map specific codes (e.g. a 401/403 pre-auth-key verdict on
 * registration) to friendly messages. Mirrors iOS `ManagementClientError.httpStatus`.
 */
class ManagementClientException(val statusCode: Int, val responseBody: String) :
  IllegalStateException(
    if (responseBody.isBlank()) "Management server returned HTTP $statusCode" else "Management server returned HTTP $statusCode: $responseBody"
  )

class ManagementClient(private val secrets: Secrets) {
  data class RegistrationResult(val clientId: String, val status: String)

  /**
   * The client record returned by `PATCH`/`GET /clients/me`, narrowed to the two fields this app acts on. [reindexPending] is `true` when the profile
   * change voided the client's queue standing; see [ProfileRefreshService] for why this client does not wait on it.
   */
  data class ClientProfile(val status: String, val reindexPending: Boolean)

  /**
   * Outcome of a conditional `GET` (`If-None-Match`). [body] is the raw response body on `200`, or `null` on `304` (the caller keeps its cached
   * copy); [etag] is the server's content hash to echo back as `If-None-Match` next time. Mirrors iOS `ManagementClient.ConditionalGet`.
   */
  data class ConditionalGet(val body: String?, val etag: String?)

  /**
   * Fetch the benchmark catalog (definitions only — no eval samples), ETag-conditional. The `GET /benchmarks` endpoints are public — no client id or
   * signature — so this needs no registration.
   */
  fun fetchBenchmarks(serverUrl: String, ifNoneMatch: String?): ConditionalGet = conditionalGet(serverUrl, "/benchmarks", ifNoneMatch)

  /**
   * Fetch a single benchmark by id, ETag-conditional. Unlike the list, the per-id response includes the eval `samples`, so every benchmark must be
   * fetched individually to be fully hydrated. Public — no auth.
   */
  fun fetchBenchmark(serverUrl: String, benchmarkId: String, ifNoneMatch: String?): ConditionalGet =
    // Percent-encode the id as a single path segment — ids come from server JSON, so an id with reserved characters must not reshape the URL path.
    conditionalGet(serverUrl, "/benchmarks/${Uri.encode(benchmarkId)}", ifNoneMatch)

  fun register(
    serverUrl: String,
    organization: String,
    contactEmail: String,
    clientDetails: String,
    publicKeyHex: String,
    preauthKey: String? = null,
  ): RegistrationResult {
    val body =
      JSONObject()
        .put("public_key", publicKeyHex)
        .put("organization", organization)
        .put("contact_email", contactEmail)
        .put("client_details", clientDetails)
    // Optional pre-auth key (`preauth_{key_id}.{secret}`): a valid one makes the server auto-approve
    // this registration instead of leaving it pending manual approval. Omitted when blank.
    if (!preauthKey.isNullOrBlank()) body.put("preauth_key", preauthKey)
    val response = request(serverUrl, "/clients/register", "POST", null, body.toString())
    val json = JSONObject(response)
    return RegistrationResult(clientId = json.getString("client_id"), status = json.getString("status"))
  }

  /**
   * `PATCH /clients/me` — refresh the device profile and capability set the planner matches jobs against. Authenticated like any other client call.
   *
   * [profile] is the request body; see [ProfileRefreshService] for which fields the server's schema accepts. `PATCH` is safe over [HttpURLConnection]
   * on Android (its implementation is OkHttp-backed, and OkHttp's method list includes `PATCH`) even though the desktop JVM's own implementation
   * rejects it.
   */
  fun updateMe(serverUrl: String, clientId: String, profile: JSONObject): ClientProfile {
    val response = request(serverUrl, "/clients/me", "PATCH", clientId, profile.toString())
    val json = JSONObject(response)
    return ClientProfile(
      status = json.getString("status"),
      // Defaulted rather than required so an older server that predates the field still parses, matching the CLI's `#[serde(default)]` and iOS's
      // `decodeIfPresent ?? false`.
      reindexPending = json.optBoolean("reindex_pending", false),
    )
  }

  fun submitResult(serverUrl: String, clientId: String, payloadJson: String): String =
    request(serverUrl, "/benchmarks", "POST", clientId, payloadJson)

  fun submitResultBatch(serverUrl: String, clientId: String, payloads: JSONArray): String {
    val body = JSONObject().put("submissions", payloads)
    return request(serverUrl, "/benchmarks/batch", "POST", clientId, body.toString())
  }

  private fun request(serverUrl: String, path: String, method: String, clientId: String?, body: String?): String {
    val url = URL(endpoint(serverUrl, path))
    val connection = url.openConnection() as HttpURLConnection
    connection.requestMethod = method
    // The `v1` signature covers the request target, so a followed redirect would present a signature over the pre-redirect target and earn a 401 —
    // and OkHttp forwards custom headers, `X-Signature` included, to the redirect host. Surface the 3xx instead.
    connection.instanceFollowRedirects = false
    connection.connectTimeout = 15_000
    connection.readTimeout = 120_000
    connection.setRequestProperty("Accept", "application/json")
    connection.setRequestProperty("User-Agent", "pipette-android")
    if (clientId != null) {
      val timestamp = DateTimeFormatter.ISO_INSTANT.format(Instant.now())
      val nonce = generateNonce()
      connection.setRequestProperty("X-Client-Id", clientId)
      connection.setRequestProperty("X-Timestamp", timestamp)
      connection.setRequestProperty("X-Nonce", nonce)
      connection.setRequestProperty("X-Signature", secrets.sign(signedPayload(method, url.requestTarget(), timestamp, clientId, nonce)))
    }
    if (body != null) {
      connection.doOutput = true
      connection.setRequestProperty("Content-Type", "application/json")
      connection.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
    }

    val status = connection.responseCode
    val stream = if (status in 200..299) connection.inputStream else connection.errorStream
    val text = stream?.bufferedReader()?.use { it.readText() } ?: ""
    if (status !in 200..299) {
      throw ManagementClientException(status, text)
    }
    return text
  }

  /**
   * Unauthenticated `GET` that honors `If-None-Match`: sends the last-seen [ifNoneMatch] etag, and a `304` returns a [ConditionalGet] with a null
   * body so the caller keeps its cached copy. Used for the public `/benchmarks` endpoints. Kept separate from [request] because that helper treats a
   * `304` as a non-success status and throws.
   */
  private fun conditionalGet(serverUrl: String, path: String, ifNoneMatch: String?): ConditionalGet {
    val connection = URL(endpoint(serverUrl, path)).openConnection() as HttpURLConnection
    connection.requestMethod = "GET"
    connection.connectTimeout = 15_000
    connection.readTimeout = 120_000
    connection.setRequestProperty("Accept", "application/json")
    connection.setRequestProperty("User-Agent", "pipette-android")
    if (ifNoneMatch != null) connection.setRequestProperty("If-None-Match", ifNoneMatch)

    val status = connection.responseCode
    val etag = connection.getHeaderField("ETag")
    if (status == HttpURLConnection.HTTP_NOT_MODIFIED) return ConditionalGet(body = null, etag = etag)

    val stream = if (status in 200..299) connection.inputStream else connection.errorStream
    val text = stream?.bufferedReader()?.use { it.readText() } ?: ""
    if (status !in 200..299) {
      throw IllegalStateException(if (text.isBlank()) "Management server returned HTTP $status" else "Management server returned HTTP $status: $text")
    }
    return ConditionalGet(body = text, etag = etag)
  }
}

/**
 * The `v1` signed payload: six newline-separated fields — scheme tag, method, request target, timestamp, client id, nonce (mgmt `authentication.md`
 * §2.1). Binding the method and target scopes a signature to that method and target; the nonce makes it single-use, so a captured signature cannot be
 * replayed inside the freshness window. The request body is still not covered.
 */
internal fun signedPayload(method: String, pathAndQuery: String, timestamp: String, clientId: String, nonce: String): String =
  "v1\n$method\n$pathAndQuery\n$timestamp\n$clientId\n$nonce"

/** Shared because [SecureRandom] is thread-safe and reseeding per request buys nothing. */
private val nonceRandom = SecureRandom()

/** 128 bits: enough that a collision across the fleet stays negligible without the client tracking what it has already sent. */
private const val NONCE_BYTES = 16

/**
 * A fresh per-request nonce: 16 CSPRNG bytes, lowercase hex.
 *
 * Hex rather than an arbitrary byte string on purpose. The nonce is a field in a newline-delimited payload, so a value carrying a newline could forge
 * a field boundary and make two different requests hash to one payload; hex cannot. It also satisfies the server's non-empty and valid-UTF-8 rules by
 * construction. 128 bits makes a collision across the fleet negligible, so the server's replay cache can reject a repeat without the client
 * coordinating.
 */
internal fun generateNonce(): String {
  val bytes = ByteArray(NONCE_BYTES)
  nonceRandom.nextBytes(bytes)
  return bytes.joinToString("") { "%02x".format(it) }
}

/**
 * Assemble an absolute endpoint URL, rejecting any server URL that would not reach the wire verbatim.
 *
 * [HttpURLConnection] is OkHttp-backed and canonicalizes the target it sends — percent-encoding characters that are not legal in a URL and resolving
 * `.`/`..` segments — while [requestTarget] reads the path back off the parsed URL as written. A server URL needing either rewrite would sign one
 * target and transmit another, 401ing every authenticated request, so it fails here instead. Three complementary guards cover what OkHttp rewrites:
 * printable-ASCII catches spaces, control characters, and non-ASCII ([URI] permits non-ASCII by documented deviation from RFC 2396, so it cannot);
 * [URI] catches the remaining illegal punctuation (`<`, `>`, `{`, `}`, `|`, `\`, `^`, `"`); and the dot-segment check reads the *decoded* path,
 * because OkHttp resolves `%2e%2e` as `..` too.
 */
internal fun endpoint(serverUrl: String, path: String): String {
  val trimmed = serverUrl.trim().trimEnd('/')
  require(trimmed.startsWith("http://") || trimmed.startsWith("https://")) { "Invalid management server URL: $serverUrl" }
  val endpoint = trimmed + path
  // `'!'..'~'` is printable ASCII minus the space — the only characters that survive to the wire unescaped.
  require(endpoint.all { it in '!'..'~' }) { "Invalid management server URL: $serverUrl" }
  val parsed =
    try {
      URI(endpoint)
    } catch (error: URISyntaxException) {
      throw IllegalArgumentException("Invalid management server URL: $serverUrl", error)
    }
  require((parsed.path ?: "").split('/').none { it == "." || it == ".." }) { "Invalid management server URL: $serverUrl" }
  return endpoint
}

/**
 * The request target the server signs over: this URL's path plus its query. Taken off the parsed URL rather than the endpoint path, so it carries any
 * path prefix the configured server URL contributes — the server verifies against the target it received.
 */
internal fun URL.requestTarget(): String = if (query == null) path else "$path?$query"

class RegistrationService(
  private val storage: LocalStorage,
  private val secrets: Secrets,
  private val client: ManagementClient,
  private val analytics: Analytics = NoOpAnalytics,
) {
  fun register(
    serverUrl: String,
    organization: String,
    contactEmail: String,
    preauthKey: String? = null,
    clerkUserId: String? = null,
    clerkSessionId: String? = null,
    clerkPrimaryEmail: String? = null,
  ): ManagementClient.RegistrationResult {
    val publicKeyHex = secrets.generatePendingSigningKeyPair()
    val result =
      try {
        client.register(
          serverUrl = serverUrl,
          organization = organization,
          contactEmail = contactEmail,
          clientDetails = DeviceInfo.modelName(),
          publicKeyHex = publicKeyHex,
          preauthKey = preauthKey,
        )
      } catch (error: Throwable) {
        secrets.deletePendingPrivateKey()
        // Coarse error kind only: a registration failure message can embed the server URL
        // and the contact email, neither of which belongs in analytics.
        analytics.capture(AnalyticsEvents.DEVICE_REGISTRATION_FAILED, mapOf(AnalyticsEvents.ERROR_KIND to AnalyticsEvents.errorKind(error)))
        throw error
      }

    if (!secrets.promotePendingPrivateKey()) {
      secrets.deletePendingPrivateKey()
      // Instrumented like the network failure above. The server accepted this device, but without a
      // usable key it can never sign a request, so leaving it uncaptured would drop the device from
      // the funnel entirely (no `device_registered`, no failure), reading as a user who wandered off
      // rather than a Keystore fault. Same reason iOS captures on its keychain-save failure.
      val error = IllegalStateException("Failed to save private key")
      analytics.capture(AnalyticsEvents.DEVICE_REGISTRATION_FAILED, mapOf(AnalyticsEvents.ERROR_KIND to AnalyticsEvents.errorKind(error)))
      throw error
    }

    // Link the Clerk identity locally at registration time (Clerk is never sent
    // to the mgmt server — registration is Ed25519-signed, identical to iOS).
    val base =
      RegistrationData(
        clientId = result.clientId,
        status = result.status,
        serverUrl = serverUrl,
        organization = organization,
        contactEmail = contactEmail,
        registeredAt = DateFormats.isoNow(),
      )
    val data =
      if (clerkUserId != null) {
        base.withClerkLink(clerkUserId, clerkSessionId, clerkPrimaryEmail)
      } else {
        base
      }
    storage.saveRegistration(data)
    // From here on every event attributes to the server-assigned device id. Deliberately no
    // email / organization / Clerk identity: the management server already holds those, and
    // PostHog has no need for them.
    analytics.identify(result.clientId)
    analytics.capture(AnalyticsEvents.DEVICE_REGISTERED, mapOf(AnalyticsEvents.STATUS to result.status))
    return result
  }
}
